pkg/cloud/services/network/routetables.go

```diff
@@ -274,28 +310,34 @@ func (s *Service) createRouteTableWithRoutes(routes []*ec2.CreateRouteInput, isP
 	record.Eventf(s.scope.InfraCluster(), "SuccessfulCreateRouteTable", "Created managed RouteTable %q", aws.ToString(out.RouteTable.RouteTableId))
 	s.scope.Info("Created route table", "route-table-id", *out.RouteTable.RouteTableId)
 
+	if err := s.createRoutesForRouteTable(routes, out.RouteTable.RouteTableId); err != nil {
+		if cleanupErr := s.deleteRouteTable(*out.RouteTable); cleanupErr != nil {
+			record.Warnf(s.scope.InfraCluster(), "FailedDeleteRouteTable", "Failed to delete managed RouteTable %q: %v", aws.ToString(out.RouteTable.RouteTableId), cleanupErr)
+		}
+		return nil, err
+	}
+
+	return &infrav1.RouteTable{
+		ID: *out.RouteTable.RouteTableId,
+	}, nil
+}
+
+func (s *Service) createRoutesForRouteTable(routes []*ec2.CreateRouteInput, routeTableID *string) error {
 	for i := range routes {
 		route := routes[i]
 		if err := wait.WaitForWithRetryable(wait.NewBackoff(), func() (bool, error) {
-			route.RouteTableId = out.RouteTable.RouteTableId
+			route.RouteTableId = routeTableID
 			if _, err := s.EC2Client.CreateRoute(context.TODO(), route); err != nil {
 				return false, err
 			}
 			return true, nil
 		}, awserrors.RouteTableNotFound, awserrors.NATGatewayNotFound, awserrors.GatewayNotFound); err != nil {
-			record.Warnf(s.scope.InfraCluster(), "FailedCreateRoute", "Failed to create route %s for RouteTable %q: %v", route, aws.ToString(out.RouteTable.RouteTableId), err)
-			errDel := s.deleteRouteTable(*out.RouteTable)
-			if errDel != nil {
-				record.Warnf(s.scope.InfraCluster(), "FailedDeleteRouteTable", "Failed to delete managed RouteTable %q: %v", aws.ToString(out.RouteTable.RouteTableId), errDel)
-			}
-			return nil, errors.Wrapf(err, "failed to create route in route table %q: %v", aws.ToString(out.RouteTable.RouteTableId), route)
+			record.Warnf(s.scope.InfraCluster(), "FailedCreateRoute", "Failed to create route %s for RouteTable %q: %v", route, aws.ToString(routeTableID), err)
```

Comment:
Here we are no longer trying to delete the route table when there was an error adding routes to it. That task has now supposed to be handled by the caller. I see that being done [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5603/commits/b5944364f5c47ff230932236b3756279b0be99d8#diff-c41ec466ac78f0428fdd8901e83494e945598aaf06958ed72e97e1b0285b9e1bR313) but not [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5603/commits/b5944364f5c47ff230932236b3756279b0be99d8#diff-c41ec466ac78f0428fdd8901e83494e945598aaf06958ed72e97e1b0285b9e1bR206). Is that intentional?

Reply:
It is currently intentionally as I thought the `fixMissingRoutes` should **not** completely remove the route table when it tries to "fix" the routes during the cluster lifecycle.

But you are right. There is a bit of discrepancies between the 2 use cases. I am looking at Marco's suggestion: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5603#discussion_r2289001346 to implement a way to handle this case.